### PR TITLE
[Wait for #2186][TFLite Export] Add Unit Test

### PR DESCRIPTION
**Pull request by DonghakPark**
## In this PR

**Add Unit Test case For Test Fused Op in Tensorflow lite export** 

Test Model consist like below

```Input -> Conv2D -> Batch Norm -> ReLU -> Flatten```

- make this model with nntrainer and export to tensorflow lite
- set input {1,3,4,4} Dim -> in Tensorflow lite native make same input with 3,4,4 and transpose this input

**latency**
(nntrainer exported has transpose layer)
10000 forward tflite native (sec)              : **0.0069193975830078125**
10000 forward nntrainter exported (sec) : **0.0070416927337646484**

**Related PR**
#2184 #2185 #2186

### CODE REVIEW

File: nntrainer/compiler/tflite_interpreter.cpp [hunk 1]
```diff
@@ -323,8 +323,13 @@ TfOpNodes buildOpNodes(const GraphRepresentation &representation,
...
     Exporter e(&fbb);
     ln->exportTo(e, ml::train::ExportMethods::METHOD_TFLITE);
+    auto export_output = e.getResult<ml::train::ExportMethods::METHOD_TFLITE>();
+
+    if (export_output.get()->getWeights().size() == 0) {
+      export_output.get()->setTrainable(false);
+    }
 
-    nodes.emplace_back(e.getResult<ml::train::ExportMethods::METHOD_TFLITE>());
+    nodes.emplace_back(move(export_output));
```
Comment (jijoongmoon) [inline nntrainer/compiler/tflite_interpreter.cpp:332]: I wonder do we need this move here? AFAIK, getResult return moved object already.

File: nntrainer/compiler/tflite_interpreter.cpp [hunk 2]
```diff
@@ -391,6 +397,73 @@ TfOpNodes buildOpNodes(const GraphRepresentation &representation,
       tf_node->weightReorder(node_count);
     }
 
+    if (tf_node->getOpType() ==
```
Comment (jijoongmoon) [inline nntrainer/compiler/tflite_interpreter.cpp:400]: Not this pr, but later, it might be better to separate these graph traversals as compiler optimizers.

File: nntrainer/compiler/tflite_interpreter.cpp [hunk 3]
```diff
@@ -391,6 +397,73 @@ TfOpNodes buildOpNodes(const GraphRepresentation &representation,
...
+          tflite::BuiltinOperator::BuiltinOperator_CONV_2D &&
+        nodes.at(node_count + 1).get()->getOpType() ==
+          tflite::BuiltinOperator::BuiltinOperator_MUL &&
+        nodes.at(node_count + 2).get()->getOpType() ==
+          tflite::BuiltinOperator::BuiltinOperator_RELU) {
```
Comment (jijoongmoon) [inline nntrainer/compiler/tflite_interpreter.cpp:405]: Is it possible to apply this fused op for sigmoid? If it is, it might be better to refactor the codes for more general cases.

File: nntrainer/compiler/tflite_interpreter.cpp [hunk 4]
```diff
@@ -622,29 +695,127 @@ buildSubGraphs(const TfOpNodes &nodes, const TfOpIdxMap &map,
 
 } // namespace
 
+TfOpNodes buildRealizedOpNodes(TfOpNodes &nodes,
```
Comment (jijoongmoon) [inline nntrainer/compiler/tflite_interpreter.cpp:698]: Is it only for the bn fusing? if it is, it is better to change the function name.

File: nntrainer/compiler/tflite_opnode.cpp [hunk 5]
```diff
@@ -111,12 +112,33 @@ void TfOpNode::setLayerNode(const LayerNode &layer) {
...
 }
 
 void TfOpNode::setWeightTransformFn(TransformFn fn) { weight_transform = fn; }
 
 void TfOpNode::setInputTransformFn(TransformFn fn) { input_transform = fn; }
 
+void TfOpNode::setWeights(Variables weights_) {
+  unsigned int cnt = 0;
+  for (auto &w : weights_) {
+    const unsigned int UNIT = w->batch();
```
Comment (jijoongmoon) [inline nntrainer/compiler/tflite_opnode.cpp:128]: let's not use capital letters for common variables.

File: nntrainer/compiler/tflite_opnode.cpp [hunk 6]
```diff
@@ -111,12 +112,33 @@ void TfOpNode::setLayerNode(const LayerNode &layer) {
...
+    const unsigned int CHANNEL = w->channel();
+    const unsigned int HEIGHT = w->height();
+    const unsigned int WIDTH = w->width();
+
+    auto weight_data = weights.at(cnt)->getData();
+    auto *ptr = const_cast<float *>(weight_data);
+    memcpy(&ptr[0], &w->getData()[0],
+           sizeof(float) * (UNIT * CHANNEL * HEIGHT * WIDTH));
+    auto weight_data2 = weights.at(cnt)->getData();
```
Comment (jijoongmoon) [inline nntrainer/compiler/tflite_opnode.cpp:137]: remove weight_data2?

File: nntrainer/compiler/tflite_opnode.cpp [hunk 7]
```diff
@@ -111,12 +112,33 @@ void TfOpNode::setLayerNode(const LayerNode &layer) {
...
+  unsigned int cnt = 0;
+  for (auto &w : weights_) {
+    const unsigned int UNIT = w->batch();
+    const unsigned int CHANNEL = w->channel();
+    const unsigned int HEIGHT = w->height();
+    const unsigned int WIDTH = w->width();
+
+    auto weight_data = weights.at(cnt)->getData();
+    auto *ptr = const_cast<float *>(weight_data);
+    memcpy(&ptr[0], &w->getData()[0],
```
Comment (jijoongmoon) [inline nntrainer/compiler/tflite_opnode.cpp:135]: how about using tensor op rather than using memcpy?

File: nntrainer/compiler/tflite_interpreter.cpp (hunk 1 above)
Comment (DonghakPark) [reply nntrainer/compiler/tflite_interpreter.cpp:332]: I think yes.
getResult move object! but for some check i already move object to export_output, so we need to move again to nodes

File: nntrainer/compiler/tflite_interpreter.cpp (hunk 2 above)
Comment (DonghakPark) [reply nntrainer/compiler/tflite_interpreter.cpp:400]: I agree with you, Thank you for Review

File: nntrainer/compiler/tflite_interpreter.cpp (hunk 3 above)
Comment (DonghakPark) [reply nntrainer/compiler/tflite_interpreter.cpp:405]: Yes. it possible and Conv + MUL + Activation is method that tflite fusing

File: nntrainer/compiler/tflite_interpreter.cpp (hunk 4 above)
Comment (DonghakPark) [reply nntrainer/compiler/tflite_interpreter.cpp:698]: Later. i will add more fusing path in this function.
it is better to change function name for now ?

File: nntrainer/compiler/tflite_opnode.cpp (hunk 5 above)
Comment (DonghakPark) [reply nntrainer/compiler/tflite_opnode.cpp:128]: FIX

File: nntrainer/compiler/tflite_opnode.cpp (hunk 6 above)
Comment (DonghakPark) [reply nntrainer/compiler/tflite_opnode.cpp:137]: FIX

File: nntrainer/compiler/tflite_opnode.cpp (hunk 7 above)
Comment (DonghakPark) [reply nntrainer/compiler/tflite_opnode.cpp:135]: can i use ```weights.at(cnt)->putData(weights_)```?

Comment (jijoongmoon) [reply nntrainer/compiler/tflite_opnode.cpp:135]: you can use https://github.com/nnstreamer/nntrainer/blob/65d5f04ab8cf666d4ec784d663a949103035d59b/nntrainer/tensor/tensor.h#L1055 
or 
https://github.com/nnstreamer/nntrainer/blob/65d5f04ab8cf666d4ec784d663a949103035d59b/nntrainer/tensor/tensor.h#L1061

**Review by baek2sm — APPROVED** [review]
LGTM

**Review by Hyoungsukim — APPROVED** [review]
lgtm

**Review by djeong20 — APPROVED** [review]
LGTM

**Review by SeoHyungjun — APPROVED** [review]
LGTM!

**Review by jijoongmoon — APPROVED** [review]
LGTM
